### ISSUE TITLE
active_support/json doesn't handle IO objects correctly

### ISSUE THREAD

**Issue by liudangyi** [rails/rails#26132]
### Steps to reproduce

```
irb(main):001:0> require 'json'
=> true
irb(main):002:0> STDOUT.to_json
=> "\"#<IO:0x007fcb1189ebd8>\""
irb(main):003:0> require 'active_support/json'
=> true
irb(main):004:0> STDOUT.to_json
IOError: not opened for reading
    from /Users/Leedy/Developer/github/rails/activesupport/lib/active_support/core_ext/object/json.rb:132:in `each'
    from /Users/Leedy/Developer/github/rails/activesupport/lib/active_support/core_ext/object/json.rb:132:in `to_a'
    from /Users/Leedy/Developer/github/rails/activesupport/lib/active_support/core_ext/object/json.rb:132:in `as_json'
    from /Users/Leedy/Developer/github/rails/activesupport/lib/active_support/json/encoding.rb:33:in `encode'
    from /Users/Leedy/Developer/github/rails/activesupport/lib/active_support/json/encoding.rb:20:in `encode'
    from /Users/Leedy/Developer/github/rails/activesupport/lib/active_support/core_ext/object/json.rb:39:in `to_json'
    from (irb):4
    from /usr/local/bin/irb:11:in `<main>'
```
### Expected behavior

`to_json` should never fail. 
### Actual behavior

It raises an error because `STDOUT.to_a` fails.

``` ruby
# activesupport/lib/active_support/core_ext/object/json.rb
module Enumerable
  def as_json(options = nil) #:nodoc:
    to_a.as_json(options)
  end
end
```
### System configuration

**Rails version**: master (5.1.0.alpha)

**Ruby version**: 2.3.1p112

**Comment by veekram** [comment]
@liudangyi I am not sure but may it has something to do with [this](https://corner.squareup.com/2014/05/ruby-serialization-enumeration.html)

**Comment by liudangyi** [comment]
@veekram Yes, I understand why it happened. But I think it's still a bug of activesupport, since `to_json` should be expected to never fail.

**Comment by sgrif** [comment]
> since to_json should be expected to never fail.

Where are you getting that idea from?

**Comment by liudangyi** [comment]
I was trying to serialize `request.env` in sinatra and found it'a OK when only requiring "json" but bugs when requiring "active_support".  

So introducing "active_support" just breaks the existing code, right?

**Comment by betesh** [comment]
This can still be reproduced on master.  Adding this to `activesupport/test/json/encoding_test.rb` demonstrates it:

```
  class Loggable
    def initialize
      @logger = ::Logger.new(STDOUT)
    end
  end

  def test_to_json_works_on_a_logger
    assert_nothing_raised { Loggable.new.to_json }
  end
```

**Comment by betesh** [comment]
This still fails on both 5-1-stable@198bc1f and master@61cc630

**Comment by betesh** [comment]
This still fails on both 5-1-stable (581fd57028) and master (70656c4453)

**Comment by rhuanbarreto** [comment]
You should fix this also for rails 5.0

**Comment by inertia186** [comment]
I just observed this in `activesupport-5.1.6`.  I have a script that does not require `activesupport`, but it's in my `Gemfile` for other scripts.  Temporarily removing `activesupport` from my `Gemfile` fixes the issue.  Problem appears to have been fixed in `activesupport-5.2.0`.
